File: csharpcore/GildedRose/ItemTypes/CheeseItemType.cs
using GildedRoseKata.Guards;

namespace GildedRoseKata.ItemTypes;

public class CheeseItemType : ItemType
{
  public CheeseItemType()
  {
    Type = "Cheese!";
  }

  public override void UpdateQuality(ref int quality, ref int sellIn)
  {
    int multiplier = sellIn < 0 ? 2 : 1;

    quality += QualityModifier * multiplier;

    quality = Guard.Against.QualityLimitations(quality);
    
    sellIn--;
  }
}

File: ruby/ItemTypes/cheese_item_type.rb
require File.join(File.dirname(__FILE__), 'item_type')
require File.join(File.dirname(__FILE__), '../guard_against')

class CheeseItemType < ItemType
  def initialize
    @type = "Cheese!"
    super
  end

  def update_quality(quality, sell_in)
    multiplier = sell_in < 0 ? 2 : 1

    quality += @quality_modifier * multiplier

    quality = GuardAgainst.quality_limitations(quality)

    sell_in -= 1

    [quality, sell_in]
  end
end
